rust-src/ed25519_hd_key_derivation/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+use hmac::{Hmac, Mac};
+use regex::Regex;
+use sha2::Sha512;
+
+const ED25519_CURVE: &[u8; 12] = b"ed25519 seed";
+const HARDENED_OFFSET: u32 = 0x80000000;
+
+pub struct HdKeys {
+    pub key:    [u8; 32],
+    chain_code: [u8; 32],
+}
+
+fn get_master_key_from_seed(seed: &[u8]) -> HdKeys {
+    let mut mac = Hmac::<Sha512>::new_from_slice(ED25519_CURVE).unwrap();
+    mac.update(&seed);
+    let i = mac.finalize().into_bytes();
+    let mut il = [0u8; 32];
+    il.clone_from_slice(&i[0..32]);
+    let mut ir = [0u8; 32];
+    ir.clone_from_slice(&i[32..64]);
+
+    HdKeys {
+        key:        il,
+        chain_code: ir,
+    }
+}
+
+fn ckd_priv(keys: HdKeys, index: u32) -> HdKeys {
+    if index & HARDENED_OFFSET == 0 {
+        panic!("Received a non-hardened index!");
+    }
+
+    let mut data = vec![0u8];
+    data.extend_from_slice(&keys.key);
+    data.extend_from_slice(&index.to_be_bytes());
+
+    let mut mac = Hmac::<Sha512>::new_from_slice(&keys.chain_code).unwrap();
+    mac.update(&data);
+    let i = mac.finalize().into_bytes();
+    let mut il = [0u8; 32];
+    il.clone_from_slice(&i[0..32]);
+    let mut ir = [0u8; 32];
+    ir.clone_from_slice(&i[32..64]);
+
+    HdKeys {
+        key:        il,
+        chain_code: ir,
+    }
+}
+
+fn is_valid_path(path: &str) -> bool {
+    let path_regex = Regex::new("^m(/[0-9]+')+$").unwrap();
+    if !path_regex.is_match(path) {
+        return false;
+    }
+
+    let segments: Vec<&str> = path.split('/').collect();
+    segments
+        .iter()
+        .skip(1)
+        .map(|seg| seg.replace("'", ""))
+        .all(|seg| seg.parse::<u32>().is_ok())
+}
+
+/// Derives hierarchical deterministic keys for ed25519 according to the SLIP0010 (https://github.com/satoshilabs/slips/blob/master/slip-0010.md)
+/// specification.
+pub fn derive(path: &str, seed: &[u8]) -> Result<HdKeys, String> {
+    if !is_valid_path(path) {
+        return Err(String::from("Invalid derivation path"));
+    }
+
+    let master_key = get_master_key_from_seed(&seed);
+
+    let segments: Vec<&str> = path.split('/').collect();
+    let segments = segments
+        .iter()
+        .skip(1)
+        .map(|seg| seg.replace("'", ""))
+        .map(|seg| seg.parse::<u32>().unwrap())
```

Comment:
If I understand correctly here (and below on lines 82/83), if I have paths `m/01'` and `m/1'` will lead to the same `HdKeys`. Is that the intention?

Reply:
That's a good point. The indices are meant to be integers, so 01 = 1 and I guess it's intentional. Maybe we don't want to allow 01 in valid paths to avoid this confusion. Maybe it would actually be better to have the type of the path be a list of integers instead of a string.

Reply:
Yes, that would be useful and cleaner (see also my comment below about constructing a string and then decomposing it just a bit later).

It sounds to me like paths are standard, so they should still be supported. But it would be nice to have an API based on Iter<Item=u32> or maybe just &[u32].

Reply:
That is correctly understood, and I guess just a bit too lenient to have it implicitly mean the same. I have updated the regex to disallow indices with prefixed `0`'s.